### PR TITLE
Add documentation for toolchain feature

### LINKED ISSUE #14225: Document toolchain feature
- [x] Release Notes
- [x] Documentation

### PR DESCRIPTION
**Pull request by bmuskalla**
Fixes #14225

### CODE REVIEW

**Comment by bmuskalla** [comment]
This sits on top of the sample PR for the snippets

File: subprojects/docs/build.gradle [hunk 1]
```diff
@@ -228,6 +228,42 @@ samples {
...
+                from(templates.javaJunit5IntegrationTestForApplication)
+
+                from(templates.javaListLibraryAsSubproject)
+                from(templates.javaJunit5TestForListLibrary)
+                from(templates.javaJunit5IntegrationTestForUtilitiesLibrary)
+
+                from(templates.javaUtilitiesLibraryAsSubproject)
+            }
+        }
+        jvmMultiProjectWithToolchains {
```
Comment (ljacomet) [inline subprojects/docs/build.gradle:249]: Duplicate or required by the sample setup?

File: subprojects/docs/src/docs/userguide/jvm/toolchains.adoc [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+[[toolchains]]
+= Toolchains for JVM projects
+
+By default, Gradle uses the same Java version for running Gradle itself and building JVM projects.
+
+This is not always desirable.
+Building projects with different Java versions on different developer machines and CI servers may lead to unexpected issues Additionally, you may want to build a project using a Java version that running Gradle is not compatible with.
```
Comment (ljacomet) [inline subprojects/docs/src/docs/userguide/jvm/toolchains.adoc:21]: Most likely missing a "." before "Additionally"

Comment (ljacomet) [reply subprojects/docs/src/docs/userguide/jvm/toolchains.adoc:21]: End of the sentence could also be reworded:

> ... a Java version that is not supported for running Gradle.

File: subprojects/docs/src/docs/userguide/jvm/toolchains.adoc [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+
+A Java Toolchain (from now on referred to simply as toolchain) is a set of tools, usually taken from a local JRE/JDK installation that are used to configure different aspects of a build.
+Compile tasks may use `javac` as their compiler, test and exec tasks may use the `java` command while
+`javadoc` will be used to generate documentation.
```
Comment (ljacomet) [inline subprojects/docs/src/docs/userguide/jvm/toolchains.adoc:25]: Extra line break

File: subprojects/docs/src/docs/userguide/jvm/toolchains.adoc [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+
+== Consuming Toolchains
```
Comment (ljacomet) [inline subprojects/docs/src/docs/userguide/jvm/toolchains.adoc:27]: Level 2 headers should have a specific link ID

File: subprojects/docs/src/docs/userguide/jvm/toolchains.adoc [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+
+====
+include::sample[dir="samples/java/jvm-multi-project-with-toolchains/groovy/",files="buildSrc/src/main/groovy/myproject.java-conventions.gradle[tags=toolchain]"]
+include::sample[dir="samples/java/jvm-multi-project-with-toolchains/kotlin/",files="buildSrc/src/main/kotlin/myproject.java-conventions.gradle.kts[tags=toolchain]"]
+====
+
+Executing the build (e.g. using `gradle check`) will now handle several things for you and others running your build
+
+1. Setup all compile and test tasks to use the defined toolchain and not the JVM that Gradle is running with
+2. Gradle detects <<sec:auto_detection,locally installed JVMs>>
```
Comment (ljacomet) [inline subprojects/docs/src/docs/userguide/jvm/toolchains.adoc:39]: Missing leading # in link

File: subprojects/docs/src/docs/userguide/jvm/toolchains.adoc [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+3. Gradle chooses a JRE/JDK matching the requirements of the build (in this case a JVM supporting Java 14)
+4. If no matching JVM is found, it will automatically download a matching JDK from https://adoptopenjdk.net/[AdoptOpenJDK]
+
+== Specify custom toolchains for individual tasks
+
+In case you want to tweak which toolchain is used for a specific task, you can specify the exact tool a task is using.
+For example, a `Test` task exposes a `JavaLauncher` property that defines which java executable to use to launch the tests.
+
+In the example below, we configure the default toolchain to use JDK8 for compiling and testing.
+Additionally, we introduce a new Test task that is going to run our unit tests but using a JDK 14. Depending on the task, a JRE might be enough while for other tasks (e.g. compilation), a JDK is required.
```
Comment (ljacomet) [inline subprojects/docs/src/docs/userguide/jvm/toolchains.adoc:49]: Missing line break on "."

Comment (ljacomet) [reply subprojects/docs/src/docs/userguide/jvm/toolchains.adoc:49]: I would also move the code extract after this first sentence and have the rest of the text below it.

File: subprojects/docs/src/docs/userguide/jvm/toolchains.adoc [hunk 7]
```diff
@@ -0,0 +1,131 @@
...
+If no toolchain is found matching the requirements of the build, Gradle will automatically download a matching JDK from AdoptOpenJDK. By default, it will request a HotSpot JDK matching the current operating system and architecture.
+Provisioning JDKs are installed in the <<directory_layout.adoc#dir:gradle_user_home,Gradle User Home directory>>.
+
+By default, the public https://api.adoptopenjdk.net/[AdoptOpenJDK APIs] are used to determine and download a matching JDK.
+In case you want to use another server that is compatible with v3 of the AdoptOpenJDK API, you can point Gradle to use a different host.
+
+```
+org.gradle.jvm.toolchain.install.adoptopenjdk.baseUri=https://api.company.net/
+```
+
```
Comment (ljacomet) [inline subprojects/docs/src/docs/userguide/jvm/toolchains.adoc:101]: Need to add the info about opting out of auto provisioning

File: subprojects/docs/build.gradle (hunk 1 above)
Comment (jjohannes) [inline subprojects/docs/build.gradle:249]: Looks like a duplicated entry (?)

**Review by ljacomet — CHANGES_REQUESTED** [review]
A couple things need to be addressed, but that's already a good start.

And yes, I wrote parts of it 😛

File: subprojects/docs/build.gradle (hunk 1 above)
Comment (ljacomet) [reply subprojects/docs/build.gradle:249]: Was wondering the same, will clean up

File: subprojects/docs/src/docs/userguide/jvm/toolchains.adoc [hunk 8]
```diff
@@ -0,0 +1,131 @@
...
+* https://sdkman.io/[SDKMAN!]
+
+=== How to disable auto-detection
+
+In order to disable auto-detection, you can use the `org.gradle.java.installations.auto-detect` gradle property.
+Either start gradle using `-Porg.gradle.java.installations.auto-detect=false` or put
+`org.gradle.java.installations.auto-detect=false` into your `gradle.properties` file.
+
+[[sec:provisioning]]
+== Auto Provisioning
```
Comment (melix) [inline subprojects/docs/src/docs/userguide/jvm/toolchains.adoc:90]: :o: Can auto provisioning be disabled? I can imagine environments where you want to fail if a compatible toolchain is not installed locally.

File: subprojects/docs/src/docs/userguide/jvm/toolchains.adoc [hunk 9]
```diff
@@ -0,0 +1,131 @@
...
+
+```
+org.gradle.java.installations.paths=/custom/path/jdk1.8,/shared/jre11
+```
+
+If your setup already provides environment variables pointing to installed JVMs, you can also let Gradle know about which environment variables to take into account.
+Assuming the environment variables `JDK8` and `JRE14` point to valid java installations, the following instructs Gradle to resolve those environment variables and consider those installations when looking for a matching toolchain.
+
+```
+org.gradle.java.installations.fromEnv=JDK8,JRE14
```
Comment (melix) [inline subprojects/docs/src/docs/userguide/jvm/toolchains.adoc:118]: :+1: I'd move this up as I think this is safer than a property which would obviously be OS specific.

File: subprojects/docs/src/docs/userguide/jvm/toolchains.adoc [hunk 10]
```diff
@@ -0,0 +1,131 @@
...
+```
+
+== Toolchains for plugin authors
```
Comment (melix) [inline subprojects/docs/src/docs/userguide/jvm/toolchains.adoc:121]: :o: Not sure I'd document this at this stage.

**Review by melix — CHANGES_REQUESTED** [review]

File: subprojects/docs/src/docs/userguide/jvm/toolchains.adoc [hunk 11]
```diff
@@ -0,0 +1,131 @@
...
+Additionally, we introduce a new Test task that is going to run our unit tests but using a JDK 14. Depending on the task, a JRE might be enough while for other tasks (e.g. compilation), a JDK is required.
+By default, Gradle prefers installed JDKs over JREs.
+
+Toolchains tool providers can be obtained from the link:{javadocPath}/org/gradle/jvm/toolchain/JavaToolchainService.html[`javaToolchains`] extension.
+
+Three tools are available:
+
+* A `JavaCompiler` which is the tool used by the `JavaCompile` task
+* A `JavaLauncher` which is the tool used by the `JavaExec` or `Test` tasks
+* A `JavadocTool` which is the tool used by the `Javadoc` task
```
[the remark this answers is not in the document]
Comment (ljacomet) [reply subprojects/docs/src/docs/userguide/jvm/toolchains.adoc:66]: Correct, will fix as we do it indeed.

File: subprojects/docs/src/docs/userguide/jvm/toolchains.adoc (hunk 8 above)
Comment (ljacomet) [reply subprojects/docs/src/docs/userguide/jvm/toolchains.adoc:95]: Yes, added already ... though the PR still needs to be merged 😉

File: subprojects/docs/src/docs/userguide/jvm/toolchains.adoc (hunk 10 above)
Comment (ljacomet) [reply subprojects/docs/src/docs/userguide/jvm/toolchains.adoc:137]: Why not?
It is part of the delivered feature. We worked to limit internal APIs usage in our own toolchain usage.

File: subprojects/docs/src/docs/userguide/jvm/toolchains.adoc (hunk 9 above)
Comment (ljacomet) [reply subprojects/docs/src/docs/userguide/jvm/toolchains.adoc:133]: Good point, inverting the two sections

File: subprojects/docs/src/docs/userguide/jvm/toolchains.adoc [hunk 12]
```diff
@@ -0,0 +1,131 @@
...
+A build can globally define what toolchain it targets by stating the Java Language version it needs:
+
+====
+include::sample[dir="samples/java/jvm-multi-project-with-toolchains/groovy/",files="buildSrc/src/main/groovy/myproject.java-conventions.gradle[tags=toolchain]"]
+include::sample[dir="samples/java/jvm-multi-project-with-toolchains/kotlin/",files="buildSrc/src/main/kotlin/myproject.java-conventions.gradle.kts[tags=toolchain]"]
+====
+
+Executing the build (e.g. using `gradle check`) will now handle several things for you and others running your build
+
+1. Setup all compile and test tasks to use the defined toolchain and not the JVM that Gradle is running with
```
[the remark this answers is not in the document]
Comment (ljacomet) [reply subprojects/docs/src/docs/userguide/jvm/toolchains.adoc:39]: Why do you want to see source and target compatibility mentioned?
Unless they have a value set by the user, we ignore them.

**Comment by ljacomet** [comment]
@bot-gradle test RFM

**Comment by ljacomet** [comment]
@bot-gradle test RFM

**Pull request closed without merging**